Complete Set Variable feature

andxu opened the pull request.

andxu committed:
add bin format

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
change IProviderAdapter to IProviderContext

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
add source path for SourceLookUpProvider

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
add file header

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
add type check in provider context impl.

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
remove local debug code

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
set variable feature.

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
remove local debug code

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
fix a jdi bug.

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
fix a style error.

Signed-off-by: andxu <andxu@microsoft.com>

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +988,188 @@ public void recyclableThreads(ThreadReference thread) {
+
+        }
+
+        private Value setValueProxy(Type type, String value, SetValueFunction setValueFunc, Map<String, Object> options)
```
No need of `SetValueFunction` class. Use `java.util.function.Consumer<Value>` as a Lambda argument.

andxu replied:
java.util.function.Consumer<Value> will force you to eat non-runtime exceptions at the lamp, SetValueFunction has the signature of throws jdi exceptions.

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +988,188 @@ public void recyclableThreads(ThreadReference thread) {
+        private Value setFieldValueWithConflict(ObjectReference obj, List<Field> fields, String name, String belongToClass,
+                                                String value, Map<String, Object> options) throws ClassNotLoadedException, InvalidTypeException {
+            Field field;
+            // first try to resolve filed by fully qualified name
```
typos `filed` -> `field`

andxu replied:
Done

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/SetValueFunction.java:
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
```
Remove this class. Use `java.util.function.Consumer` instead.

andxu replied:
use SetValueFunction to allow unchecked exception InvalidTypeException, ClassNotLoadedException

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +988,188 @@ public void recyclableThreads(ThreadReference thread) {
+            if (arguments.name.contains("(")) {
+                name = arguments.name.substring(0, arguments.name.indexOf('(')).trim();
+                belongToClass = arguments.name.substring(arguments.name.indexOf('(') + 1, arguments.name.indexOf(')'))
+                        .trim();
```
Suggest to use regex to parse the value.

andxu replied:
done

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +988,188 @@ public void recyclableThreads(ThreadReference thread) {
+                                Field field = type.fieldByName(name);
+                                newValue = setStaticFieldValue(type, field, arguments.name, arguments.value, options);
+                            } else {
+                                if (frame.location().method().isStatic() && showStaticVariables) {
```
`if ` in this line `1030` is unnecessary. It duplicates with parent if `1024`

andxu replied:
NICE catch.

andxu committed:
remove SettingProvideraa

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
1. fix the typo filed
2. use regex to extract a pattern like 'a (b)'
3. remove a duplicate condition.

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
1. remove useless param from VariableRequestHandler@ctor
2. fix checkstyle issue.

Signed-off-by: andxu <andxu@microsoft.com>

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/JdiObjectProxy.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.util.Objects;
+
+public class JdiObjectProxy<T> {
```
What's the purpose of this one?
Didn't see special logic other than the members already existed on an Object.

andxu replied:
the jdi impl calls the valid state function in hashcode, if we resume a thread, the related stackframe will throw invalid state exception when evaluating its hashcode, but we need the hashcode of the stackframe to remove the stackframe, so I create a wrapper object:
    private final Map<V, Set<O>> referenceMap = new HashMap<>();
    private final Map<V, Integer> objectIdMap = new HashMap<>();
...
public boolean removeObjectById(int id) {
        synchronized (this) {
            V object = this.objectCollection.remove(id);
            if (object == null)  {
                return false;
            }
            referenceMap.remove(object);
            objectIdMap.remove(object);
            return true;
        }
    }

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/IVariableFormatter.java:
```diff
@@ -38,6 +38,15 @@
+     * @return the matched value formatter, if no formatter matches,
+     * <code>IllegalArgumentException</code> will be thrown
+     */
+    IValueFormatter getValueFormatter(Type type, Map<String, Object> options);
```
I think our original idea was to only expose valueToString and stringToValue. The proper formatter should be determined by passing options. Why do we need to expose this?

andxu replied:
valuetostring is designed for formatting an existing jdi value,  typetostring is designed for formatting the type of a jdi value, we can get a value formatter from a jdi value, and type formatter from a jdi type, on setVariable scenario, we need a value formatter from a jdi type to construct a jdi value, so the getValueFormatter(other another proper name) is required in VariableFormatter, we can remove the valuetostring, but we cannot avoid the usage of 
getValueFormatter

andxu replied:
added public Value stringToValue(String stringValue, Type type, Map<String, Object> options)  to VariableFormatter

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/VariableUtils.java:
```diff
@@ -142,10 +143,20 @@ public static boolean hasChildren(Value value, boolean includeStatic) {
-                Variable var = new Variable("arg" + argId, argValue);
-                var.argumentIndex = argId++;
-                res.add(var);
+            try {
```
What's this about?

andxu replied:
it is a jdi bug which throws InteralException on getArgumentValues:

https://community.oracle.com/thread/1523023

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+        Responses.ResponseBody setVariable(Requests.SetVariableArguments arguments) {
+            Map<String, Object> options = new HashMap<>();
+            // TODO: when vscode protocol support customize settings of value format, showQualified should be one of the options.
+            boolean showStaticVariables = true;
```
This should be false by default. User will need to explicitly turn it on by configuring launch.json. If your purpose is to increase test coverage, then I agree that we can keep this value until we the official release.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+            Map<String, Object> options = new HashMap<>();
+            // TODO: when vscode protocol support customize settings of value format, showQualified should be one of the options.
+            boolean showStaticVariables = true;
+            boolean showQualified = true;
```
showQualified -> showFullyQualifiedNames
The current name is vague.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+            if (newValue instanceof ArrayReference) {
+                indexedVariables = ((ArrayReference) newValue).length();
+            }
+            return new Responses.SetVariablesResponseBody(
```
This function is just toooooooooooooooooooooooo long.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                throws ClassNotLoadedException, InvalidTypeException {
+            IValueFormatter formatter = getFormatterForModification(type, options);
+            Value newValue = formatter.valueOf(value, type, options);
+            setValueFunc.apply(newValue);
```
Why do we need a callback to handle the set value scenario? Can it be another member function of this class?

andxu replied:
Can it be another member function of this class? yes, it can be.

There will be a various of setValue functions, stackFrame.setValue, array.setValue,  object.setValue,  classType.setValue, all of them are using different parameters and throwing same exceptions, using a SetValueFunction as the signature to keep these code reused but kepts there exceptions, otherwise you need to catch the InvalidTypeException, ClassNotLoadedException exceptions, but have no clean way of reporting to the callers.  

Object:void setValue(Field field, Value value)
        throws InvalidTypeException, ClassNotLoadedException;
Array:   void setValue(int index, Value value)
            throws InvalidTypeException,
                   ClassNotLoadedException;
StackFrame:void setValue(LocalVariable variable, Value value)
        throws InvalidTypeException, ClassNotLoadedException;
ClassType: void setValue(Field field, Value value) throws InvalidTypeException, ClassNotLoadedException

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                belongToClass = arguments.name.replaceFirst(PATTERN, "$2");
+            }
+            try {
+                if (obj instanceof StackFrameScope) {
```
This branch should be in a new function.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                                    String.format("SetVariableRequest: Variable %s cannot be found.", arguments.name));
+                        }
+                    }
+                } else if (obj instanceof ThreadObjectReference) {
```
This branch should also be a new function.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
         }
 
+        Responses.ResponseBody setVariable(Requests.SetVariableArguments arguments) {
+            Map<String, Object> options = new HashMap<>();
```
I think the correct way to initialize the "options" is to call VariableFormatter.getDefaultOptions(). Then later we can change its values.

andxu replied:
Done

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                name = arguments.name.replaceFirst(PATTERN, "$1");
+                belongToClass = arguments.name.replaceFirst(PATTERN, "$2");
+            }
+            try {
```
A blank line should be added before the try block.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+            return newValue;
+        }
+
+        private IValueFormatter getFormatterForModification(Type type, Map<String, Object> options) {
```
This function should be long to VariableFormatter. By passing proper "options", it should be able to determine the right formatter to work with the type.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                throw new UnsupportedOperationException(
+                        String.format("SetVariableRequest: Field %s in interface cannot be changed.", name));
+            }
+            return setValueProxy(field.type(), value, newValue -> ((ClassType) declaringType).setValue(field, newValue), options);
```
Now I understand the use of the proxy. But it's not actually a proxy, it's the real code that sets the value.

akaroml replied:
For the exception issue of the Consumer interface, I don't see you throw any exception in the lambda function. Can we still use Consumer for easier understanding?

andxu replied:
interface SetValueFunction {
        void apply(Value value) throws InvalidTypeException, ClassNotLoadedException;
    }
because SetValueFunction is able to defining the throws clause, so I don't need to catch these exceptions, just to delegate these exceptions to the callee.

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:
```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+            return setValueProxy(eleType, value, newValue -> array.setValue(index, newValue), options);
+        }
+
+        private Value setFieldValueWithConflict(ObjectReference obj, List<Field> fields, String name, String belongToClass,
```
Please help understand the use scenarios of this function.

andxu replied:
The name conflict can have the following scenarios:
1. the name conflict happens, but each name belongs to a different class name(after formatted).
2.  the name conflict happens, and unfortunately the  belonged to class name still conflicts, eg:
//Dup.java
package dup;
public class Dup {
    public  int x = 2;
}

//DupTest.java
package dup2;

class Dup extends dup.Dup {
    public  int x = 1;
}
public class DupTest extends Dup {
    public int x = 3;
}

andxu replied:
In the second scenario, we always format the type name to full qualified name.

andxu committed:
1. change SetValueFunction to inner class of DebugAdapter
2. create the jdi value from formatter

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
1. split functions in a big method setVariable

Signed-off-by: andxu <andxu@microsoft.com>

andxu committed:
1. fix a bug when terminating debug session multiple times, a connection error error raises.

Signed-off-by: andxu <andxu@microsoft.com>

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/VariableFormatter.java:
```diff
@@ -92,6 +92,12 @@ public String valueToString(Value value, Map<String, Object> options) {
     }
 
+    @Override
+    public Value stringToValue(String stringValue, Type type, Map<String, Object> options) {
```
Perfect!

akaroml commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/variables/IVariableFormatter.java:
```diff
@@ -53,6 +53,15 @@
+     * @param options additional information about expected format
+     * @return the jdi value
+     */
+    Value stringToValue(String stringValue, Type type, Map<String, Object> options);
```
Perfect!

akaroml reviewed (dismissed):
Thanks!

andxu committed:
1. remove debug code.

Signed-off-by: andxu <andxu@microsoft.com>

testforstephen commented on org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/DebugSession.java:
```diff
@@ -60,7 +60,9 @@ public void detach() {
     @Override
     public void terminate() {
-        vm.exit(0);
+        if (vm.process().isAlive()) {
```
check vm.process() is null or not.

testforstephen replied:
the debug session doesn't need check vm is alive or not. Leave the check to the caller.
